cluster/cluster.go

```diff
@@ -0,0 +1,127 @@
+package cluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/groupcache/groupcache-go/v3"
+	"github.com/groupcache/groupcache-go/v3/transport"
+	"github.com/groupcache/groupcache-go/v3/transport/peer"
+)
+
+var _daemons []*Daemon
+var _peers []peer.Info
+
+// ListPeers returns a list of all peers in the cluster
+func ListPeers() []peer.Info {
+	return _peers
+}
+
+// ListDaemons returns a list of all daemons in the cluster
+func ListDaemons() []*Daemon {
+	return _daemons
+}
+
+// DaemonAt returns a specific daemon
+func DaemonAt(idx int) *Daemon {
+	return _daemons[idx]
+}
+
+// PeerAt returns a specific peer
+func PeerAt(idx int) peer.Info {
+	return _peers[idx]
+}
+
+// FindOwningDaemon finds the daemon which owns the key provided
+func FindOwningDaemon(key string) *Daemon {
+	c, isRemote := _daemons[0].GroupCache.PickPeer(key)
+	if !isRemote {
+		return _daemons[0]
+	}
+
+	for i, d := range _daemons {
+		if d.ListenAddress() == c.PeerInfo().Address {
+			return _daemons[i]
+		}
+	}
+	panic(fmt.Sprintf("failed to find daemon which owns '%s'", key))
+}
+
+// Start a local cluster
+func Start(ctx context.Context, numInstances int, opts groupcache.Options) error {
+	var peers []peer.Info
+	port := 1111
+	for i := 0; i < numInstances; i++ {
+		peers = append(peers, peer.Info{
+			Address: fmt.Sprintf("localhost:%d", port),
+		})
+		port += 1
+	}
+	return StartWith(ctx, peers, opts)
+}
+
+// StartWith a local cluster with specific addresses
+func StartWith(ctx context.Context, peers []peer.Info, opts groupcache.Options) error {
+	if len(_daemons) != 0 || len(_peers) != 0 {
+		return errors.New("StartWith: cluster already running; shutdown the previous cluster")
```

Comment:
We can wrap this error into an error variable. That can help an application to kind of use the error type for some validation.

Reply:
Unless there is a current use case for creating a custom error type, we should not create something no one is using, in the hopes that one day it will be useful.